### PR TITLE
feat(multiple-prop): added the ability to select multiple values

### PR DESCRIPTION
**Pull request by souporserious**
**What**:
Related to https://github.com/paypal/react-kadabra/issues/61

**Why**:
So we can build more complex menus that allow selecting multiple values.

**How**:
I followed how the browser handles `multiple` and `value`. When `multiple` is `true`, we handle value as an array. I switched the API from `selectedItem` -> `value` because I think this is a little more clear since you are technically selecting the value of your item[s]. This is also how other select menus like react-select handle this so it seemed like a safe change. I changed `value` -> `inputValue` which I think is more clear as well.

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged 
- [ ] Added myself to contributors table 

I was able to build a select menu, a multi-select, an autocomplete, and a simple dropdown menu with these changes. It's pretty cool how powerful, but simple it is! I _think_ this is closing in on a pretty final API after this. But please, let me know what you think 😇  if all looks well, I'll update the rest of the examples. It's a pretty minimal breaking change. Mainly `value` -> `inputValue`.

### CODE REVIEW

File: src/autocomplete.js [hunk 1]
```diff
@@ -8,18 +8,18 @@ import {cbToCb, composeEventHandlers, debounce, scrollIntoView} from './utils'
...
     getA11yStatusMessage: PropTypes.func,
     getValue: PropTypes.func,
+    multiple: PropTypes.bool,
     onChange: PropTypes.func.isRequired,
     onClick: PropTypes.func,
+    value: PropTypes.any,
   }
 
   static defaultProps = {
-    defaultSelectedItem: null,
```
Comment (kentcdodds) [inline src/autocomplete.js]: Should this be `defaultValue: null`? Even though it doesn't actually make a difference, I like having defaults set for anything that's not required.

File: src/autocomplete.js [hunk 2]
```diff
@@ -164,38 +179,43 @@ class Autocomplete extends Component {
...
+      toggleMenu,
       selectItem,
       selectItemAtIndex,
       selectHighlightedItem,
       setHighlightedIndex,
-      getRootProps,
+      clearSelection,
+    } = this
+    return {
+      // prop getters
```
Comment (kentcdodds) [inline src/autocomplete.js]: Thank you!

File: src/autocomplete.js [hunk 3]
```diff
@@ -164,38 +179,43 @@ class Autocomplete extends Component {
...
-      clearSelection,
+      toggleMenu,
       selectItem,
       selectItemAtIndex,
       selectHighlightedItem,
       setHighlightedIndex,
-      getRootProps,
+      clearSelection,
+    } = this
+    return {
```
Comment (kentcdodds) [inline src/autocomplete.js]: Could we expose `getItemFromIndex` while we're at it? Ref: https://github.com/paypal/react-kadabra/issues/71

File: src/autocomplete.js [hunk 4]
```diff
@@ -288,15 +308,19 @@ class Autocomplete extends Component {
...
-    const {value, selectedItem, isOpen} = this.state
+    const {isOpen} = this.state
     return {
       'data-autocomplete-input': true,
       role: 'combobox',
       'aria-autocomplete': 'list',
       'aria-expanded': isOpen,
       autoComplete: 'off',
-      value: (value === null ? this.getValue(selectedItem) : value) || '',
+      value: this.state.inputValue,
```
Comment (kentcdodds) [inline src/autocomplete.js]: Let's destructure this above. I like to destructure stuff at the top of the closure so it's easy to tell what this function uses.

File: src/autocomplete.js [hunk 5]
```diff
@@ -308,14 +332,12 @@ class Autocomplete extends Component {
     if (event.key && this.keyDownHandlers[event.key]) {
       this.keyDownHandlers[event.key].call(this, event)
     } else if (!['Shift', 'Meta', 'Alt', 'Control'].includes(event.key)) {
-      this.openMenu(() => {
-        this.highlightIndex()
-      })
+      this.openMenu()
```
Comment (kentcdodds) [inline src/autocomplete.js]: Why this change?

File: src/autocomplete.js [hunk 6]
```diff
@@ -338,11 +360,10 @@ class Autocomplete extends Component {
   //\\\\\\\\\\\\\\\\\\\\\\\\\\\\\ ITEM
 
   reset = () => {
-    this.setState(({selectedItem}) => ({
+    this.setState({
       isOpen: false,
       highlightedIndex: null,
-      value: this.getValue(selectedItem),
```
Comment (kentcdodds) [inline src/autocomplete.js]: The `inputValue` should be reset to the `getValue` of the currently selected item right?

File: src/autocomplete.js [hunk 7]
```diff
@@ -371,10 +396,6 @@ class Autocomplete extends Component {
     this.toggleMenu(false, cb)
   }
 
-  getValue = item => {
```
Comment (kentcdodds) [inline src/autocomplete.js]: Why is this removed?

File: src/autocomplete.js (hunk 3 above)
Comment (souporserious) [inline src/autocomplete.js]: done ✅

File: src/autocomplete.js (hunk 4 above)
Comment (souporserious) [inline src/autocomplete.js]: Oops! On it.

File: src/autocomplete.js (hunk 5 above)
Comment (souporserious) [inline src/autocomplete.js]: I moved that logic into `toggleMenu` so anytime the menu is opened it will handled highlighting.

File: src/autocomplete.js (hunk 6 above)
Comment (souporserious) [inline src/autocomplete.js]: Oops, that was a mistake, putting it back 😁

Comment (souporserious) [inline src/autocomplete.js]: Ohh that wasn't a mistake. I handle that in `selectItem` now. I actually need to handle `multiple` and `inputValue`.

File: src/autocomplete.js (hunk 7 above)
Comment (souporserious) [inline src/autocomplete.js]: I moved it closer to the input methods :)

File: src/autocomplete.js (hunk 1 above)
Comment (souporserious) [inline src/autocomplete.js]: My bad. Done 👍

**Comment by souporserious** [comment]
I added some other things I found 😁 it looks like the recent highlighting changes got lost in the new API from today. I added that back, as well as the button API we had before, which turned out nice because we now have access to the same handlers as Input uses 🎉

**Comment by souporserious** [comment]
Do you think `value` should be `selectedValue` so it's a little more explicit? 🤔

**Comment by kentcdodds** [comment]
> Do you think value should be selectedValue so it's a little more explicit? 🤔

Yes 👍

File: src/autocomplete.js [hunk 8]
```diff
@@ -257,6 +285,30 @@ class Autocomplete extends Component {
 
   //////////////////////////// BUTTON
 
+  buttonKeyDownHandlers = {
+    ArrowDown: this.keyDownHandlers.ArrowDown,
+
+    ArrowUp: this.keyDownHandlers.ArrowUp,
+
+    Enter(event) {
```
Comment (kentcdodds) [inline src/autocomplete.js]: Actually, I think I prefer "Enter" on the button to select the highlighted item.

I'd love to hear what @marcysutton has to say on the subject of accessibility. Right now you can hit the spacebar on the button to toggle the menu and the up/down arrow keys to change the highlighted item, then the enter key to select an item. If we add this event handler, I feel like the Enter key has too much going on:

1. It can open the menu
2. It can close the menu
3. It can select the highlighted item.

I'd rather just have the enter key select the highlighted item if one is highlighted (like it is currently). [Here's the interaction](https://codesandbox.io/s/9rWEmzEv8) as it stands today: 

a11y

Comment (souporserious) [inline src/autocomplete.js]: Totally 👍 I did that wrong. I meant for space to handle that functionality :) that's seems to be how the default browser select works.

**Comment by souporserious** [comment]
Alright. I _think_ I have everything done. I was wondering in `onChange` whether we should pass the `selectedValue` there as well for convenience. That way you can see what value was just selected/deselected as well as look at the current state of selections.

**Comment by kentcdodds** [comment]
> I was wondering in onChange whether we should pass the selectedValue there as well for convenience.

Yes. This sounds like a good idea. Though what if an item is removed in a multi-select scenario?

**Comment by souporserious** [comment]
This is why I think we almost need to pass it now. That way you can compare what was just selected with the current values to determine that. I ended up with [something like this for selectly](https://github.com/souporserious/selectly/blob/master/src/Select.jsx#L146-L150) that worked well at our work. I don't _think_ we need to pass as much as I was passing though.

**Comment by kentcdodds** [comment]
Got it so the API would be:

`onChange(currentValue, previousValue)`?

**Comment by kentcdodds** [comment]
I kinda like what you have there actually. Maybe we should do an object...

```js
onChange({selectedValue, previousValue})
```

**Comment by souporserious** [comment]
I like that. I think that's enough info to do what you need 🤔 maybe this will be good enough for now and we can revisit this part if we need to.

**Comment by kentcdodds** [comment]
I wonder if we could also provide what kind of change it was: `onChange({selectedValue, previousValue, type: Autocomplete.changeTypes.removal})`

And in `changeTypes` we'd have `removal`, `addition`

Would that be useful? Should we concern ourselves with that?

**Comment by tannermares** [comment]
@souporserious: I'm torn on wether or not the select should be that smart too. I have needed that value a couple times but wasn't super annoyed I had to do the subtraction myself to figure it out. I'm afraid it'll start a trend of random helper props that just bloats things. I love where this is going either way!

**Comment by kentcdodds** [comment]
> I'm afraid it'll start a trend of random helper props that just bloats things

Yeah, let's skip the `type` thing and just stick with `selectedValue` and `previousValue`. We can always add it later. Harder to remove it ;-)

**Comment by souporserious** [comment]
Cool :) that's what I was thinking. It's always easier to add than remove.

**Comment by souporserious** [comment]
I landed on this. Thoughts?
image

**Comment by kentcdodds** [comment]
Cool. So `itemValue` would be the item that was just selected? What would this look like the context of a non-multiselect? What would `itemValue` be if an item is removed?

**Comment by souporserious** [comment]
Haha I just deleted that comment 😆 I think I was doing pretty much what we talked about above. Let's see how `selectedValue` and `previousValue` play out.

File: src/autocomplete.js (hunk 8 above)
Comment (marcysutton) [inline src/autocomplete.js]: Yep, space bar is the default for buttons.

**Comment by souporserious** [comment]
Alright, I _think_ it's finally done 😁

File: src/autocomplete.js [hunk 9]
```diff
@@ -137,11 +140,36 @@ class Autocomplete extends Component {
...
+          const values = [...state.selectedValue]
+          const pos = values.indexOf(itemValue)
+          if (pos > -1) {
+            values.splice(pos, 1)
+          } else {
+            values.push(itemValue)
+          }
+          return {
+            selectedValue: values,
+            inputValue: values.map(value => this.getValue(value)).join(', '),
```
Comment (kentcdodds) [inline src/autocomplete.js]: In the future I expect we'll need to make this a prop as well `getInputValue`... Either that or instead of mapping all the values to `this.getValue` we just call `this.getValue(values)`... Seems like that'd be reasonable... Should we do that now?

**Comment by kentcdodds** [comment]
Sorry, I'm going to push forward with the rename. I want to get that done to avoid confusing people longer! Hopefully there will be no merge conflicts.

File: src/autocomplete.js (hunk 9 above)
Comment (souporserious) [inline src/autocomplete.js]: Ah yes, that's true. The end user might want to handle arrays in a different way or maybe they wouldn't want to, I'm not sure 🤔

**Comment by souporserious** [comment]
Looks like no merge conflicts 🙌

File: src/autocomplete.js (hunk 9 above)
Comment (souporserious) [inline src/autocomplete.js]: Maybe ok to push through and see how people react? Keeps the API a little easier IMO

Comment (kentcdodds) [inline src/autocomplete.js]: Sounds fine.

**Comment by kentcdodds** [comment]
I'll merge this now. Release coming when I finish with my late lunch

**Comment by kentcdodds** [comment]
```
~/Developer/downshift (master)
🐑  $ npm publish --tag beta
+ downshift@1.0.0-beta.9
```
